specs/phase0/fork-choice.md

```diff
@@ -304,13 +362,57 @@ def store_target_checkpoint_state(store: Store, target: Checkpoint) -> None:
 
 ```python
 def update_latest_messages(store: Store, attesting_indices: Sequence[ValidatorIndex], attestation: Attestation) -> None:
-    target = attestation.data.target
+    attestation_slot = attestation.data.slot
     beacon_block_root = attestation.data.beacon_block_root
     for i in attesting_indices:
-        if i not in store.latest_messages or target.epoch > store.latest_messages[i].epoch:
-            store.latest_messages[i] = LatestMessage(epoch=target.epoch, root=beacon_block_root)
+        if i not in store.latest_messages or attestation_slot > store.latest_messages[i].slot:
+            node_key = get_node_key(beacon_block_root, attestation_slot)
+            store.latest_messages[i] = LatestMessage(slot=attestation_slot, root=node_key)
 ```
 
+#### `add_block_slot_node`
+
+```python
+def add_block_slot_node(store: Store, block: BeaconBlock, node_slot: Slot) -> BlockSlotNode:
+    """
+    Adds a ``BlockSlotNode`` corresponding to ``block`` at slot ``node_slot`` in ``store.block_slot_tree`` 
+    and returns the same.
+
+    Algorithm:
+        1. Check if the parent block's node exists at slot ``block.slot - 1``. 
+           If not found, recursively call ``add_block_slot_node()`` for parent block's node.
+        2. Add this block's node from ``block.slot`` till ``node_slot``. 
+    """
+    assert block.slot <= node_slot
+
+    # Check if parent block's node exists at ``block.slot-1``
+    # Check for parent block's node only if current block is not genesis/anchor block
+    # NOTE: This only checks if ``block`` is the genesis block. Arbitrary anchor block capability
+    # is not implemented yet.
+    if block.slot > GENESIS_SLOT:
+        parent_block_node_slot = Slot(block.slot - 1)
+        parent_node_key = get_node_key(block.parent_root, parent_block_node_slot)
+        # Add the parent block's node if it doesn't exist
+        if parent_node_key not in store.block_slot_tree.keys():
```

Comment:
nitpicking: it doesn't need `keys()` for checking the existence of a key in a Dict.
```suggestion
        if parent_node_key not in store.block_slot_tree:
```